src/Resources/views/includes/sidebar/_profile.html.twig

```diff
@@ -1,3 +1,4 @@
+{% macro profile(name, edit_link = null, name_detail = null, picture_src = null) %}
```

Comment:
Nit: As with components in other templating systems, we have to take architecture of a growing “component” system into consideration. I don't think there is a good, official styleguide. But here is something that I learned:

- Separating configuration options from data.
- Having a first parameter to be data, a second one to be configuration.
- Having an easy way to extend data, therefore passing it as an object.

As macros grow, having all properties as individual parameters will not be effortless to maintain, as you will have several breaking changes to introduce along the way.

Based on this, the recommendation I would have for upcoming macros is, passing the data in a first parameter as an object, e.g. `{{ profile({ name: name, edit_link: edit_link, picture_src: picture_src}) }}`. This allows to update the data object parameters without changing the macro signature.

Second, if you have settings later like additional HTML Attributes, themes etc., this signature could look like this:

```twig
{% macro profile(data, opts) %} ....
```

```twig
{{ profile({
  name: name,
  edit_link: edit_link,
  picture_src: picture_src
}, {
  variant: 'compact',
  attr: {
    class: 'example'
  }
) }}
```

I also often look at the form macros of Symfony for inspiration

Reply:
Good thinking. I changed the params to an object.

One thing which we can't do anymore is setting a default value and forcing an input. Do you know a solution to these issues? The name and link need to be set.

Reply:
@marcwieland95 

You set defaults like this, inside the macro:

```twig
{% set name = data.name|default('my-default') %}
```

Or if you want to mirror it into the incoming data object, you can do it as follows. This won't work 1:1 with nested data due to it now being shallow/deep merge. If you pass data that has nested objects, you need to manually use the merge-filter on the nested properties.

```twig
{% set default_data = {
  name: 'my-name'
} %}

{% set data = default_data|merge(data) %}
```

I recommend not to use merge on data inputs (only options/properties) and rather use the first approach. If `data` can be a PHP class instead of a raw array (which is possible), the merge-filter will throw an error. So if you assume the developers will pass PHP classes as data, do not try to use the merge filter directly on the data object.